### ISSUE TITLE
Accuracy issue of ReLU network encryption inference.

### ISSUE THREAD

**Issue by yt-candy** [TrustworthyComputing/REDsec#6]
Hello, sir! Sorry to bother you, but the accuracy of my encryption inference using relu as the activation function is very low.

Using the `relu1024x3x3.py` file to retrain the network, it was found that the training accuracy was very low. Parameter tuning revealed that the problem was with  `threshold-value=0.1`, which was modified to 0.05. The training accuracy was greatly improved.
<img width="455" alt="Image" src="https://github.com/user-attachments/assets/18be1c6d-6504-46b5-b592-947e63af9ed7" />
And  found that the parameter of relu is set to 4, but the input of pixels is 8-bit. May I ask why the parameter is set in this way?
<img width="396" alt="Image" src="https://github.com/user-attachments/assets/9dfc6773-a929-4e4c-a1cb-e5745933d960" />

I tried to use my trained network for inference, and the trained network is as follows（ `threshold-value=0.05`）：

<img width="898" alt="Image" src="https://github.com/user-attachments/assets/745baa52-6f0c-44cd-98f5-5adf21e0f07f" />

I did not modify the training data and still used MNIST data.
The training accuracy is 95%, and the make ptxt accuracy is 100% (I modified NUM_SAMPLES from 1 to 100)

<img width="648" alt="Image" src="https://github.com/user-attachments/assets/b3e00fc5-0a81-4a91-b793-a418eb6a4ed1" />

But it still cannot solve the problem of accuracy in encryption reasoning.

If it is a limitation of the relu activation function implementation or if there is an issue with my operation?

Thanks for your help!

**Comment by lfolkerts** [comment]
Hello,

For ReLU activations, there is expected modest drop in accuracy between the python and plaintext/encrypted implementations due to the discretization of BatchNorm. However, plaintext and encrypted models should match. There may be an issue with the TFHE parameterization that is causing the accuracy to deteriorate in the encrypted domain.

The reason we use 4-bit precision is to allow for efficient TFHE parameters that allow for the correct decryption of ciphertexts. Increasing precision beyond what the parameter set can handle will lead to improper results and large accuracy drops.

**Comment by yt-candy** [comment]
> Hello,
> 
> For ReLU activations, there is expected modest drop in accuracy between the python and plaintext/encrypted implementations due to the discretization of BatchNorm. However, plaintext and encrypted models should match. There may be an issue with the TFHE parameterization that is causing the accuracy to deteriorate in the encrypted domain.
> 
> The reason we use 4-bit precision is to allow for efficient TFHE parameters that allow for the correct decryption of ciphertexts. Increasing precision beyond what the parameter set can handle will lead to improper results and large accuracy drops.
Thank you for your reply!

**Comment by yt-candy** [comment]
Sorry to bother you again, I want to use the ReLU network to do some work, but the citertext reasoning accuracy of the ReLU network is surprisingly low.
Running the original code directly without any modification, the reasoning accuracy of the ReLU network (relu1024x1, 1024x2 and 1024x3) is almost only 10% to 20%. Although you have said before that the accuracy of ciphertext reasoning in the ReLU network will decrease, I think this is already a very abnormal phenomenon compared with the accuracy rate of 90% of the clear text.
So I would like to ask if there is a problem with any parameter configuration, and if you could provide some assistance, I would be very grateful！
